File: python/counting/counting.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Example of multi-threading in Python for handling counting."""

import threading
import sys

class CommonCtx:
    """Common context to be used in threads"""

    def __init__(self, total_count, do_debug = False):
        self.total_count = total_count
        self.do_debug = do_debug
        self.printer_tid = 0
        self.cur_count = 0
        self.cur_count_handled = False
        self.lock = threading.Lock()
        self.cv = threading.Condition(self.lock)

    def __repr__(self):
        return "(total_count={}, ".format(self.total_count) + \
            "do_debug={}, printer_tid={} ".format(self.do_debug,
                                                  self.printer_tid) + \
            "cur_count={}, cur_count_handled={})".format(self.cur_count,
                                                         self.cur_count_handled)

class ThreadCtx:
    """Thread-specific context to be used in threads"""

    def __init__(self, id, common_ctx):
        self.id = id
        self.common_ctx = common_ctx

    def __repr__(self):
        return "(id={}, common_ctx={})".format(self.id, self.common_ctx)

def tprint(str):
    """Utility for thread-safe printing"""

    sys.stdout.write(str + "\n")

def counting_fn(thread_ctx):
    """Thread counting function"""

    ctx = thread_ctx.common_ctx
    id = thread_ctx.id

    if (ctx.do_debug):
        tprint("starting thread {}".format(id))

    with ctx.cv:
        if (ctx.do_debug):
            tprint("(thread {}) got lock".format(id))
        while (True):
            if (id is ctx.printer_tid):
                tprint("(thread {}) cur_count={}".format(id, ctx.cur_count))
                ctx.cur_count_handled = True
                ctx.cv.notify()
            if (ctx.cur_count >= ctx.total_count):
                break
            ctx.cv.wait()

    if (ctx.do_debug):
        tprint("finishing thread {}".format(id))

THREAD_COUNT = 5
ctx = CommonCtx(50)

thread_list = []
for i in range(THREAD_COUNT):
    thread_ctx = ThreadCtx((i + 1), ctx)
    t = threading.Thread(target=counting_fn, args=(thread_ctx,))
    t.daemon = True
    t.start()
    thread_list.append((t, thread_ctx))

with ctx.cv:
    ctx.printer_tid = 0
    for ctx.cur_count in range(1, (ctx.total_count + 1)):
        ctx.printer_tid = \
            1 if (ctx.printer_tid >= THREAD_COUNT) else (ctx.printer_tid + 1)
        ctx.cur_count_handled = False
        tprint("cur_count={}".format(ctx.cur_count))
        ctx.cv.notify_all()
        while (not ctx.cur_count_handled):
            ctx.cv.wait()

try:
    for (t, t_ctx) in thread_list:
        t.join(1.0)
        if (t.is_alive()):
            tprint("thread {} did not finish".format(t_ctx.id))
except KeyboardInterrupt:
    tprint("Caught keyboard interrupt")

tprint("{} threads finished".format(THREAD_COUNT))

File: cpp/counting/src/counting.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <vector>
#include <tuple>

struct CommonCtx {
    uint8_t printerTid_;
    uint8_t curCount_;
    bool curCountHandled_;

    uint8_t totalCount_;
    bool doDebug_;

    std::mutex m_;
    std::condition_variable cv_;

    explicit CommonCtx(bool doDebug, uint8_t totalCount) :
        totalCount_(totalCount), doDebug_(doDebug) {}
    CommonCtx(const CommonCtx &) = delete;
    CommonCtx & operator=(const CommonCtx&) = delete;
    CommonCtx(CommonCtx &&) = delete;
    CommonCtx & operator=(CommonCtx&&) = delete;
};

struct ThreadCtx {
    uint8_t threadId_;
    CommonCtx &commonCtx_;

    explicit ThreadCtx(CommonCtx &ctx, uint8_t threadId) :
        threadId_(threadId), commonCtx_(ctx) {}
};

static void
countingFn(std::shared_ptr<ThreadCtx> ctx)
{
}

int
main(int argc, char **argv)
{
    const auto threadCount = 5;
    const auto totalCount = 50;
    const auto doDebug = false;

    CommonCtx ctx(doDebug, totalCount);
    //std::vector<std::tuple<std::thread, ThreadCtx>> threadVec;
    std::vector<std::thread> threadVec;

    for (auto i = 0; i < threadCount; ++i) {
        auto tCtx = std::make_shared<ThreadCtx>(ctx, (i + 1));
#if 0
        threadVec.push_back(std::make_tuple(std::thread(countingFn, tCtx),
                                            tCtx));
#endif
        threadVec.push_back(std::thread(countingFn, tCtx));
    }

    auto &tid = ctx.printerTid_;
    {
        std::unique_lock<std::mutex> lk(ctx.m_);
        tid = 0;
        for (ctx.curCount_ = 1; ctx.curCount_ <= ctx.totalCount_;
             ++(ctx.curCount_)) {
            tid = (tid >= threadCount) ?  1 : (tid + 1);
            ctx.curCountHandled_ = false;
            std::cout << __func__ << ": curCount=" << ctx.curCount_ <<
                std::endl;
            ctx.cv_.notify_all();
            ctx.cv_.wait(lk, [&]{ return ctx.curCountHandled_; });
        }
    }

    for (auto &it : threadVec) {
        //std::get<0>(it).join();
        it.join();
    }

    return 0;
}
